---
title: "Pertussis Resurgence"
author: "Vera Sophia Beliaev"
date: "3/8/2022"
output: pdf_document
---

# Investigating Pertussis Cases by Year

Web scraping from the CDC website with the help of the datapasta package: let's you copy and paste in data from a website and it will be interpretted as a data frame

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
library(datapasta)
library(ggplot2)
```

```{r}
#put data into clipboard,go to Addins then click paste as dataframe 

cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )
```

```{r}
ggplot(cdc) + aes(Year, No..Reported.Pertussis.Cases) + geom_point() + geom_line() + labs(x="Year", y= "Number of Cases", title= "Pertussis Cases by Year(1922-2019")
```


# A Tale of Two Vaccines (wP & aP)

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
ggplot(cdc) + aes(Year, No..Reported.Pertussis.Cases) + geom_point() + geom_line() + labs(x="Year", y= "Number of Cases", title= "Pertussis Cases by Year(1922-2019") +geom_vline(xintercept =1946, linetype= "dotted", color = "blue") + geom_vline(xintercept=1996, linetype= "dotted", color= "red") +geom_text(label="wP", x= 1950, y=270000, color = "blue") + geom_text(label= "aP", x= 2000, y= 270000, color= "red")
```

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?
Pertussis cases were increasing after the introduction of the aP vaccine. There could be many explanations including lower rates of vaccination, evolving Bordetella pertussi, waning immunity, more PCR testing, or less efficacy of the aP vaccine compared to the wP vaccine.

# Exploring SMI-PB Data

CMI-PB tracks "long-term humoral and cellular immune response data for a large number of individuals who received either DTwP or DTaP combination vaccines in infancy followed by Tdap booster vaccinations"

CMI-PB API gives JSON data which is formatted as a series of key-value pairs (keys/particular words are associated w/ a aprticular value)

JSON format example:
{
"isotype" : "IgG",
"antigen" : "PT"
}

to read JSON files, use read_json() fn in jsonlite package or rjson package

```{r}
#jsonlite can simplify JSON key-value pair arrays into R data frames
library(jsonlite)
```

```{r}
#Read the main subject database table, metadata about study participants
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?
There are 47 aP infancy vaccinated subjects and 49 wP infancy vaccinated subjects in the dataset.

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?
There are 66 female subjects and 30 male subjects in the dataset.

```{r}
table(subject$biological_sex)
```

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?
The breakdown of race and biological sex is listed in the table below.

```{r}
table(subject$biological_sex, subject$race)
```

Side-Note: Working with dates

```{r}
library(lubridate)
```
```{r}
today()
```

in lubridate, use ymd() to tell the format of the data's date, and use time_length( , "years) fn to convert days to years

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?
The average age of wP individuals is 35 years old and the average age of aP individuals is 24 years old.

Problem on Knit:
library(dplyr)
wp <- subject %>%
  filter(infancy_vac == "wP")
round(summary(time_length(ap$age, "years")))
ap <- subject %>%
  filter(infancy_vac == "aP")
round(summary(time_length(ap$age, "years")))

```{r}
library(dplyr)
```

> Q8. Determine the age of all individuals at time of boost?
The average age of individuals at the time of boost in 26 years old.

```{r}
#new col w/ age of individs at time of boost in days
subject$boostage <-ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
```


```{r}
round(summary(time_length(subject$boostage, "years")))
```


> Q8. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?
Yes, based on the faceted boxplot, the two groups are significantly different.

```{r}
subject$age <- today() - ymd(subject$year_of_birth)
```


```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```



# Joining Multiple Tables

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```

We need to link/join/merge the specimen and subject data frames with dplyr's join() fn


> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
dim(subject)
```


```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```


> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
#dim() shows # of rows and cols
abdata <- inner_join(titer, meta)
dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?
The number of specimens for each isotype is specified in the table below.

```{r}
table(abdata$isotype)
```

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?
There are way less number of 8 visit specimens compared to the other visits.

```{r}
table(abdata$visit)
```

# Examine IgG1 Titer Levels

```{r}
#filter for IgG1 isotype, excluding the visit 8 entries
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

> Q13.  Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) + aes(ab_titer, antigen) + geom_boxplot() + facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?
Antigens such as PT, FIM2/3, and FHA show differences in the level of IgG1 antiody titers recognizing them over time. These may be antigens whose expression changes over time.

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen == "FIM2/3") %>%
  ggplot() + aes(ab_titer, col=infancy_vac) + geom_boxplot(show.legend = TRUE) + facet_wrap(vars(visit)) + theme_bw() + labs(title= "FIM2/3 antigen levels per visit")
```

```{r}
filter(ig1, antigen == "FHA") %>%
  ggplot() + aes(ab_titer, col=infancy_vac) + geom_boxplot(show.legend = TRUE) + facet_wrap(vars(visit)) + theme_bw() + labs(title= "FHA antigen levels per visit")
```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?
The antigens appear to increase over time for the first few visits then decrease over time for the last few visits.

> Q17. Do you see any clear difference in aP vs. wP responses?
In the case of FIM 2/3 antigen, wP vaccianted appear to have slightly higher antigen levels compared to aP vaccinated. In the case of FHA antigen, aP vaccinated appear to have less variance in antigen levels than wP vaccianted.

# Obtaining CMI-PB RNASeq Data

We will obtain RNA-Seq results for specific ENSEMBLE gene identifiers which can be combined with the & character

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE)
```

"Join" the rna expression data to our metadata named meta

```{r}
ssrna <- inner_join(rna, meta)
```

> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
# Note: tpm is TPM expression values
ggplot(ssrna) + aes(visit, tpm, group=subject_id) + geom_point() + geom_line(alpha=0.2)
```
> Q19. What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?
The expression of the IGHG1 gene increases then decreases back down over time. It is at its maximum expression when measured at the fourth visit.

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?
No because previously it appeared that antigens reach their peak levels when measured at the fifth visit. This may be because antibodies persist longer in the body than the expression of an individual gene.


No obvious aP vs wP difference in expression of IGHG1 gene (shown below).
```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```
Below, we focus on a single visit to evaluate whether there is wP vs aP difference in IGHG1 expression however there is no obvious difference.

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

# Working with Larger Datasets

We will use the "2020 longitudinal RNA-Seq data" from CMI-PB, provided as a CSV file

```{r}
# Change for your downloaded file path
rnaseq <- read.csv("/Users/belia/Downloads/2020LD_rnaseq.csv")

head(rnaseq,3)
```

```{r}
# number of rows and cols in this huge dataset
dim(rnaseq)
```

## Working with Long Format Data

The rnadata is in "long format" rather than the conventional wide format
In wide format, rows=genes, cols= counts in different experiments

```{r}
#Check how many genes are reported for e/ specimen_id
n_genes <- table(rnaseq$specimen_id)
head(n_genes, 10)
```

```{r}
# Check the number of specimens
length(n_genes)
```

```{r}
#Check that the # of genes is the same for all speciments
all(n_genes[1]==n_genes)
```

### Convert to "Wide" Format

use pivot_wider() fn from tidyr package

```{r}
library(tidyr)
```

```{r}
rna_wide <- rnaseq %>%
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)
dim(rna_wide)
```

```{r}
head(rna_wide[,1:7], 3)
```













